File: backend/dao/repositories/inference.py
from __future__ import annotations

import json
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from dao.models.inference import InferenceEvent, InferenceRequestRecord


class InferenceRepository:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def create(
        self,
        *,
        request_id: UUID,
        model: str,
        status: str,
        decision: str,
        scan_id: UUID | None,
        user_id: UUID | None,
        worker_id: str | None,
        end_to_end_latency_ms: float,
        security_overhead_ms: float,
        inference_latency_ms: float | None,
        output_preview: str | None,
        metadata: dict[str, object] | None = None,
    ) -> InferenceRequestRecord:
        record = InferenceRequestRecord(
            request_id=request_id,
            user_id=user_id,
            scan_id=scan_id,
            model=model,
            status=status,
            decision=decision,
            worker_id=worker_id,
            end_to_end_latency_ms=end_to_end_latency_ms,
            security_overhead_ms=security_overhead_ms,
            inference_latency_ms=inference_latency_ms,
            output_preview=output_preview,
            metadata_json=json.dumps(metadata or {}, default=str),
        )
        self.session.add(record)
        await self.session.flush()
        return record

    async def add_event(
        self,
        inference_id: UUID,
        event_type: str,
        payload: dict[str, object] | None = None,
    ) -> InferenceEvent:
        event = InferenceEvent(
            inference_id=inference_id,
            event_type=event_type,
            payload_json=json.dumps(payload or {}, default=str),
        )
        self.session.add(event)
        await self.session.flush()
        return event

    async def get_by_request_id(self, request_id: UUID) -> InferenceRequestRecord | None:
        result = await self.session.execute(
            select(InferenceRequestRecord).where(InferenceRequestRecord.request_id == request_id)
        )
        return result.scalar_one_or_none()

    async def list_recent(self, *, limit: int = 50) -> list[InferenceRequestRecord]:
        result = await self.session.execute(
            select(InferenceRequestRecord)
            .order_by(InferenceRequestRecord.created_at.desc())
            .limit(limit)
        )
        return list(result.scalars().all())

File: backend/dao/store/inference.go
package store

import (
	"context"
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
)

type InferenceRecord struct {
	ID                   uuid.UUID
	RequestID            uuid.UUID
	UserID               *uuid.UUID
	ScanID               *uuid.UUID
	Model                string
	Status               string
	Decision             string
	WorkerID             *string
	EndToEndLatencyMs    float64
	SecurityOverheadMs   float64
	InferenceLatencyMs   *float64
	OutputPreview        *string
	MetadataJSON         string
	CreatedAt            time.Time
}

type InferenceRepo struct {
	pool *pgxpool.Pool
}

func NewInferenceRepo(pool *pgxpool.Pool) *InferenceRepo {
	return &InferenceRepo{pool: pool}
}

type CreateInferenceParams struct {
	RequestID           uuid.UUID
	Model               string
	Status              string
	Decision            string
	ScanID              *uuid.UUID
	UserID              *uuid.UUID
	WorkerID            *string
	EndToEndLatencyMs   float64
	SecurityOverheadMs  float64
	InferenceLatencyMs  *float64
	OutputPreview       *string
	Metadata            map[string]interface{}
}

func (r *InferenceRepo) Create(ctx context.Context, p CreateInferenceParams) (*InferenceRecord, error) {
	now := time.Now().UTC()
	id := uuid.New()
	meta := p.Metadata
	if meta == nil {
		meta = map[string]interface{}{}
	}
	metaJSON, _ := json.Marshal(meta)
	_, err := r.pool.Exec(ctx, `
		INSERT INTO inference_requests (id, request_id, user_id, scan_id, model, status, decision,
			worker_id, end_to_end_latency_ms, security_overhead_ms, inference_latency_ms,
			output_preview, metadata_json, created_at, updated_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15)`,
		id, p.RequestID, p.UserID, p.ScanID, p.Model, p.Status, p.Decision,
		p.WorkerID, p.EndToEndLatencyMs, p.SecurityOverheadMs, p.InferenceLatencyMs,
		p.OutputPreview, string(metaJSON), now, now)
	if err != nil {
		return nil, err
	}
	return &InferenceRecord{
		ID: id, RequestID: p.RequestID, UserID: p.UserID, ScanID: p.ScanID,
		Model: p.Model, Status: p.Status, Decision: p.Decision, WorkerID: p.WorkerID,
		EndToEndLatencyMs: p.EndToEndLatencyMs, SecurityOverheadMs: p.SecurityOverheadMs,
		InferenceLatencyMs: p.InferenceLatencyMs, OutputPreview: p.OutputPreview,
		MetadataJSON: string(metaJSON), CreatedAt: now,
	}, nil
}

func (r *InferenceRepo) AddEvent(ctx context.Context, inferenceID uuid.UUID, eventType string, payload map[string]interface{}) error {
	if payload == nil {
		payload = map[string]interface{}{}
	}
	data, _ := json.Marshal(payload)
	now := time.Now().UTC()
	_, err := r.pool.Exec(ctx, `
		INSERT INTO inference_events (id, inference_id, event_type, payload_json, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6)`,
		uuid.New(), inferenceID, eventType, string(data), now, now)
	return err
}

func (r *InferenceRepo) ListRecent(ctx context.Context, limit int) ([]InferenceRecord, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT id, request_id, user_id, scan_id, model, status, decision, worker_id,
			end_to_end_latency_ms, security_overhead_ms, inference_latency_ms, output_preview,
			metadata_json, created_at
		FROM inference_requests ORDER BY created_at DESC LIMIT $1`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []InferenceRecord
	for rows.Next() {
		var rec InferenceRecord
		if err := rows.Scan(&rec.ID, &rec.RequestID, &rec.UserID, &rec.ScanID, &rec.Model,
			&rec.Status, &rec.Decision, &rec.WorkerID, &rec.EndToEndLatencyMs,
			&rec.SecurityOverheadMs, &rec.InferenceLatencyMs, &rec.OutputPreview,
			&rec.MetadataJSON, &rec.CreatedAt); err != nil {
			return nil, err
		}
		out = append(out, rec)
	}
	return out, rows.Err()
}

func (r *InferenceRepo) AvgSecurityOverhead(ctx context.Context) (float64, error) {
	var avg float64
	err := r.pool.QueryRow(ctx, `SELECT COALESCE(AVG(security_overhead_ms), 0) FROM inference_requests`).Scan(&avg)
	return avg, err
}

func (r *InferenceRepo) P95DetectionLatency(ctx context.Context) (float64, error) {
	var p95 float64
	err := r.pool.QueryRow(ctx, `
		SELECT COALESCE(percentile_cont(0.95) WITHIN GROUP (ORDER BY latency_ms), 0)
		FROM security_scans`).Scan(&p95)
	return p95, err
}
